src/zenml/logging/step_logging.py

```diff
@@ -126,25 +224,41 @@ def write(self, text: str) -> None:
 
         if not self.disabled:
             self.buffer.append(text)
+            self.save_to_file()
 
-            if (
-                len(self.buffer) >= self.max_messages
-                or time.time() - self.last_save_time >= self.time_interval
-            ):
-                self.save_to_file()
+    @property
+    def _is_write_needed(self) -> bool:
+        """Checks whether the buffer needs to be written to disk.
 
-    def save_to_file(self) -> None:
-        """Method to save the buffer to the given URI."""
-        if not self.disabled:
+        Returns:
+            whether the buffer needs to be written to disk.
+        """
+        return (
+            len(self.buffer) >= self.max_messages
+            or time.time() - self.last_save_time >= self.time_interval
+        )
+
+    def save_to_file(self, force: bool = False) -> None:
+        """Method to save the buffer to the given URI.
+
+        Args:
+            force: whether to force a save even if the write conditions not met.
+        """
+        if not self.disabled and (self._is_write_needed or force):
             # IMPORTANT: keep this as the first code line in this method! The
             # code that follows might still emit logging messages, which will
             # end up triggering this method again, causing an infinite loop.
             self.disabled = True
 
-            artifact_store = Client().active_stack.artifact_store
             try:
                 if self.buffer:
-                    with artifact_store.open(self.logs_uri, "a") as file:
+                    with self.artifact_store.open(
+                        os.path.join(
+                            self.logs_uri_folder,
+                            f"{time.time()}{LOGS_EXTENSION}",
+                        ),
```

Comment:
This multi-file approach will create a new file/object every 15 seconds. For a 24 hours running step, this results in roughly 5700 files/objects being stored, which becomes another scaling problem.

To ensure better scalability, the number of log files in the artifact store must not grow linearly with the step's runtime duration. I'll think of a solution for this.

Reply:
Please, keep in mind immutable files on GCS, since it was the main driver for the multifile approach.
We can also roll up files more often, rather than on step exit only.

Reply:
And not necessarily every 15 seconds - if no new logs arrive, the save should not be called until then.